nx/lib/nx/random.ex

```diff
@@ -0,0 +1,390 @@
+    count_size = Nx.axis_size(count, 0)
+    even? = rem(count_size, 2) == 0
+
+    if even? do
+      count
+    else
+      Nx.concatenate([count, Nx.tensor([0])])
+    end
+  end
+
+  defp threefry2x32_out_transform(count, original) do
+    count_size = Nx.axis_size(original, 0)
+    even? = rem(count_size, 2) == 0
+
+    if even? do
+      count
+    else
+      count
+      |> Nx.flatten()
+      |> Nx.slice_along_axis(0, count_size, axis: 0)
+    end
+  end
+
+  defnp threefry2x32_20(xs, ks) do
+    rotations = Nx.tensor([[13, 15, 26, 6], [17, 29, 16, 24]], type: {:u, 8})
+    key1 = ks[0]
+    key2 = ks[1]
+
+    xs = Nx.add(ks, xs)
+
+    ks =
+      Nx.stack([
+        key2,
+        Nx.bitwise_xor(key1, key2)
+        |> Nx.bitwise_xor(0x1BD11BDA),
+        key1
+      ])
+      |> Nx.as_type({:u, 32})
+
+    state = {xs, ks, rotations}
+
+    {_, {nxs, _, _}} =
+      while {x = 0, state}, Nx.less(x, 5) do
+        {x + 1, rolled_loop_step(x, state)}
+      end
+
+    nxs
+  end
+
+  defnp apply_round(xs, rot) do
+    y1 = Nx.add(xs[0], xs[1])
+
+    y2 =
+      rotate_left(xs[1], rot)
+      |> Nx.bitwise_xor(y1)
+
+    Nx.stack([y1, y2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  defnp rolled_loop_step(i, {xs, ks, rs}) do
+    {k1, k2, k3} = {ks[0], ks[1], ks[2]}
+    {r1, r2} = {rs[0], rs[1]}
+
+    {_, xs, _} =
+      while {x = 0, xs, rs}, Nx.less(x, 4) do
+        {x + 1, apply_round(xs, rs[0][x]), rs}
+      end
+
+    xs1 =
+      Nx.broadcast(k1, xs[0])
+      |> Nx.add(xs[0])
+
+    xs2 =
+      Nx.broadcast(k2 + i + 1, xs[1])
+      |> Nx.add(xs[1])
+
+    new_xs =
+      Nx.stack([xs1, xs2])
+      |> Nx.as_type({:u, 32})
+
+    new_ks = Nx.stack([k2, k3, k1])
+    new_rs = Nx.stack([r2, r1])
+
+    {new_xs, new_ks, new_rs}
+  end
+
+  defnp rotate_left(x, rot) do
+    nbits = 32
+    x <<< rot ||| x >>> (nbits - rot)
+  end
+
+  defp finfo(type) do
+    case type do
+      {:bf, 16} -> [exp: 8, mant: 7]
+      {:f, 16} -> [exp: 5, mant: 11]
+      {:f, 32} -> [exp: 8, mant: 23]
+      {:f, 64} -> [exp: 11, mant: 52]
+    end
+  end
+
+  @doc """
+  Sample uniform random integer values in [minval, maxval).
+  
+  ## Options
+  
+    * `:type` - an int type for the returned tensor
+  
+    * `:shape` - shape of the returned tensor
+  
+  """
+  defn randint(key, minval, maxval, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:s, 32})
+    shape = opts[:shape]
+    type = {_, nbits} = opts[:type]
+
+    transform(key, &randint_transform(&1, type))
+
+    keys = transform(key, &split/1)
+
+    minval = Nx.broadcast(minval, shape)
+    maxval = Nx.broadcast(maxval, shape)
+
+    higher_bits = transform(keys[0], &random_bits(&1, shape))
+    lower_bits = transform(keys[1], &random_bits(&1, shape))
+    span = maxval - minval
+
+    multiplier =
+      Nx.power(2, Nx.quotient(nbits, 2))
+      |> Nx.remainder(span)
+      |> Nx.power(2)
+      |> Nx.remainder(span)
+
+    offset =
+      higher_bits

```

Comment:
same with the offset, it would be nice to add a comment on where this calculation comes from (or a general reference for both at the top of the function)